delocate/libsana.py

```diff
@@ -48,73 +226,94 @@ def tree_libs(start_path, filt_func=None):
-            search_paths = rpaths + env_var_paths
-            for install_name in get_install_names(depending_libpath):
-                # If the library starts with '@rpath' we'll try and resolve it
-                # We'll do nothing to other '@'-paths
-                # Otherwise we'll search for the library using env variables
-                if install_name.startswith('@rpath'):
-                    lib_path = resolve_rpath(install_name, search_paths)
-                elif install_name.startswith('@'):
-                    lib_path = install_name
-                else:
-                    lib_path = search_environment_for_lib(install_name)
-                if lib_path in lib_dict:
-                    lib_dict[lib_path][depending_libpath] = install_name
-                else:
-                    lib_dict[lib_path] = {depending_libpath: install_name}
+            for dependancy_path, install_name in get_dependencies(
+                depending_path
+            ):
+                if install_name.startswith("@loader_path/"):
+                    # Support for `@loader_path` would break existing callers.
+                    logger.debug(
+                        "Excluding %s because it has '@loader_path'.",
+                        install_name
+                    )
+                    continue
+                lib_dict.setdefault(dependancy_path, {})
+                lib_dict[dependancy_path][depending_path] = install_name
     return lib_dict
 
 
-def resolve_rpath(lib_path, rpaths):
-    """ Return `lib_path` with its `@rpath` resolved
-
-    If the `lib_path` doesn't have `@rpath` then it's returned as is.
+def resolve_rpath(lib_path, rpaths, loader_path, executable_path="."):
+    # type: (Text, Iterable[Text], Text, Text) -> Text
+    """ Return `lib_path` with any special runtime linking names resolved.
 
     If `lib_path` has `@rpath` then returns the first `rpaths`/`lib_path`
-    combination found.  If the library can't be found in `rpaths` then a
-    detailed warning is printed and `lib_path` is returned as is.
+    combination found.  If the library can't be found in `rpaths` then
+    DependencyNotFound is raised.
+
+    `@loader_path` and `@executable_path` are resolved with their respective
+    parameters.
 
     Parameters
     ----------
     lib_path : str
-        The path to a library file, which may or may not start with `@rpath`.
+        The path to a library file, which may or may not be a relative path
+        starting with `@rpath`, `@loader_path`, or `@executable_path`.
     rpaths : sequence of str
         A sequence of search paths, usually gotten from a call to `get_rpaths`.
+    loader_path : str
+        The path to be used for `@loader_path`.
+        This must be the directory of the library which is loading `lib_path`.
+    executable_path : str, optional
+        The path to be used for `@executable_path`.
+        Assumed to be the working directory by default.
```

Comment:
It isn't a good default. It was hard to imagine what circumstances would lead to having to delocate a Python library using `@executable_path`, but I put something there anyway.

I can use `sys.base_exec_prefix` to make `@executable_path` relative to the Python executable. That seems like a reasonable choice.

Reply:
`dirname(sys.executable)` is now the default for `@executable_path`, and this can be changed from the command line.